### ISSUE TITLE
Mapsforge maps v5: custom tag keys

### ISSUE THREAD

**Issue by b3nn0** [mapsforge/mapsforge#1041]
Hi
Since File format version 5, we now finally have the ability to place arbitrary values into the map file, only by providing a key. The value will then be stored in the map file.

The next step - which should be quite simple to achieve - would be to also allow these arbitrary values as captions.
Right now, TextKey.java will only accept these keys:
```
    private static final String KEY_ELEVATION = "ele";
    private static final String KEY_HOUSENUMBER = "addr:housenumber";
    private static final String KEY_NAME = "name";
    private static final String KEY_REF = "ref";
```
If any other key is to be used for caption, it will do a
` throw new IllegalArgumentException("invalid key: " + key);`

However, since the tags are already there, it would be very easy to use arbitrary keys as caption,
simply by, instead of throwing an exception, returning new TextKey(key).
Or, if we wan't to be a bit smarter and not do too many allocations, using a static map, mapping a key string to TextKey instances.

**Comment by devemux86** [comment]
[Map file format specification](https://github.com/mapsforge/mapsforge/blob/master/docs/Specification-Binary-Map-File.md) defines specific space in poi and way data for those fields.

**Comment by b3nn0** [comment]
Yes, I know. But isn't this - more or less - due to historic reasons?
I don't see any reason why it shouldn't be allowed to use arbitrary keys - everything is there and the changes would be minor, without breaking backward compatibility.

I'm using mapsforge for non-OSM data ("Flurstücke"/"Plots").
My usecase would be: show a short label when zoomed out, show the full plot number when zooming in.
I could, of course make these labels show up by abusing e.g. the "name" and "ref" keys, but that is kinda ugly and there is really no technical reason for it.

Would a patch be accepted that allows for arbitrary caption keys to be used in RenderTheme?
If so, would you prefer a e.g. a static HashMap containing the TextKey instances, or rather creating new TextKey instances for each label?

**Comment by devemux86** [comment]
I have created many proprietary maps with existing fields without issues.

Still not understand what want to do here.
Option to replace the "ref" with "xxxxx" field or support more fields in map file?

**Comment by b3nn0** [comment]
No, just basically replace
`throw new IllegalArgumentException("invalid key: " + key);`
by
`return new TextKey(key);`
or something equivalent.

That's all that's needed to allow RenderTheme rules like
`  <caption k="myOwnKey" /> `
instead of being limited to ele/addr:housnumber/name/ref.

It is surely not absolutely required, as long as 4 possible captions are enough for a single Node/Way, bit it surely is nicer, and more generic.

**Comment by devemux86** [comment]
That's in client.
What about map files and writer?
Have you written map files with "myOwnKey" and read them correctly?

**Comment by b3nn0** [comment]
Yes, that works fine, since map format version 5.
E.g. assume the following tag-values file:
```
        <osm-tag key="name"    value="%s" zoom-appear="14"/>
        <osm-tag key="flstkz"    value="%s" zoom-appear="14"/>
        <osm-tag key="fst_nr_c"  value="%s" zoom-appear="14"/>
        <osm-tag key="ownerType" value="%s" zoom-appear="14"/>
```
With this, all the Strings will be in the map file (I can see it in the debugger).
However, I can only use `<caption k="name"> ` and not `<caption k="flstkz" />`
However, when checking out the rendering process, I can see that the tags are there:
![grafik](https://user-images.githubusercontent.com/1858945/37112358-7a8ecd58-2242-11e8-8e5e-40287c727a9a.png)

I just can't reference them in the RenderTheme because of the exception.

**Comment by devemux86** [comment]
Ah now I see what you mean, if you want can post a pull request.
We need to make good test with all these, memory efficient is always welcome. :slightly_smiling_face:

**Comment by devemux86** [comment]
Fixed via #1042.

**Comment by devemux86** [comment]
[VTM](https://github.com/mapsforge/vtm) client already supports custom tag keys with Mapsforge maps v5.
